plugins/rts54hub/fu-rts54hub-rtd21xx-foreground.c

```diff
@@ -0,0 +1,374 @@
+		return FALSE;
+	}
+
+	/* wait for device ready */
+	g_usleep(300000);
+	if (!fu_rts54hub_rtd21xx_device_i2c_read (FU_RTS54HUB_RTD21XX_DEVICE (self),
+						  UC_ISP_SLAVE_ADDR,
+						  0x00,
+						  buf_rep, sizeof(buf_rep),
+						  error)) {
+		g_prefix_error (error, "failed to get version number: ");
+		return FALSE;
+	}
+	/* set version */
+	version = g_strdup_printf ("%u.%u", buf_rep[1], buf_rep[2]);
+	fu_device_set_version (FU_DEVICE (self), version);
+
+	if (!fu_rts54hub_rtd21xx_device_i2c_write (FU_RTS54HUB_RTD21XX_DEVICE (self),
+						   UC_ISP_SLAVE_ADDR,
+						   UC_FOREGROUND_OPCODE,
+						   buf, sizeof(buf),
+						   error)) {
+		g_prefix_error (error, "failed to ISP_CMD_FW_UPDATE_RESET: ");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_foreground_detach_raw (FuRts54hubRtd21xxForeground *self, GError **error)
+{
+	guint8 buf = 0x03;
+	if (!fu_rts54hub_rtd21xx_device_i2c_write (FU_RTS54HUB_RTD21XX_DEVICE (self), 0x6A, 0x31,
+						   &buf, sizeof(buf), error)) {
+		g_prefix_error (error, "failed to detach: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_foreground_detach_cb (FuDevice *device,
+					  gpointer user_data,
+					  GError **error)
+{
+	FuRts54hubRtd21xxForeground *self = FU_RTS54HUB_RTD21XX_FOREGROUND (device);
+	guint8 status = 0xfe;
+
+	if (!fu_rts54hub_rtd21xx_foreground_detach_raw (self, error))
+		return FALSE;
+	if (!fu_rts54hub_rtd21xx_device_read_status_raw (FU_RTS54HUB_RTD21XX_DEVICE (self),
+							 &status, error))
+		return FALSE;
+	if (status != ISP_STATUS_IDLE_SUCCESS) {
+		g_set_error (error,
+				FWUPD_ERROR,
+				FWUPD_ERROR_INTERNAL,
+				"detach status was 0x%02x", status);
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_foreground_detach (FuDevice *device, GError **error)
+{
+	FuRts54HubDevice *parent = FU_RTS54HUB_DEVICE (fu_device_get_parent (device));
+	g_autoptr(FuDeviceLocker) locker = NULL;
+
+	/* open device */
+	locker = fu_device_locker_new (parent, error);
+	if (locker == NULL)
+		return FALSE;
+	return fu_device_retry (device,
+				fu_rts54hub_rtd21xx_foreground_detach_cb,
+				100, NULL, error);
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_foreground_attach (FuDevice *device, GError **error)
+{
+	FuRts54HubDevice *parent = FU_RTS54HUB_DEVICE (fu_device_get_parent (device));
+	FuRts54hubRtd21xxForeground *self = FU_RTS54HUB_RTD21XX_FOREGROUND (device);
+	guint8 write_buf[ISP_PACKET_SIZE] = { 0x0 };
+	g_autoptr(FuDeviceLocker) locker = NULL;
+
+	/* open device */
+	locker = fu_device_locker_new (parent, error);
+	if (locker == NULL)
+		return FALSE;
+
+	/* exit fw mode */
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_RESTART);
+	fu_device_set_progress (device, 0);
+	if (!fu_rts54hub_rtd21xx_device_read_status (FU_RTS54HUB_RTD21XX_DEVICE (self),
+						     NULL, error))
+		return FALSE;
+	write_buf[0] = ISP_CMD_FW_UPDATE_EXIT;
```

Comment:
for read version， should Send "ISP_CMD_FW_UPDATE_RESET " in attach instead of in fu_rts54hub_rtd21xx_ensure_version_unlocked  funtion.

Reply:
Agree, fixed, thanks.